Refactor validation tests to use compiletest

XAMPPRocky opened the pull request:
This PR moves nearly all validation tests out of `spirv-builder` and moves them into `tests/` which is a crate that uses completest to compile and build each each test. I've also added the `cargo testsuite` command which is a short alias to run the tests with `--release` mode, as it is substantially quicker (~40s in debug vs ~7s in release). You do need to run these tests in a separate command, as including it in the same workspace was leading to issues with conflicting files from cargo. Overall this implementation speeds up testing by quite a bit on my local machine.

Disassembly tests are still to be done, as I haven't figured how to get those working with SPIR-V yet, but that can be done in a follow-up PR.

closes #429

khyperia commented on tests/ui/push_constant_vulkan.rs:
```diff
@@ -0,0 +1,26 @@
+// (`#[spirv(block)]` here) on `struct ShaderConstants`.
+
+// build-pass
+// compile-flags: -Ctarget-feature=+vulkan
```
I don't think this runs vulkan validation, it only enables the vulkan memory model.

XAMPPRocky replied:
Some we add a target feature to run Vulkan validation?

khyperia commented on tests/build.rs:
```diff
@@ -0,0 +1 @@
+fn main() {}
```
Why is this needed?

XAMPPRocky replied:
Needed in an older version, but I forgot to delete it.

khyperia commented on Cargo.toml:
```diff
@@ -1,4 +1,6 @@
+
```
nit: extra line

XAMPPRocky commented on tests/ui/allocate_const_scalar_pointer.rs:
```diff
@@ -0,0 +1,17 @@
+// Doesn't work, only worked before because I think it got optimized away before
+// hitting the backend.
```
Hmm yeah, I can replicate it failing, when building as a shader, but it is passing in the testsuite. I'll have to investigate more.

XAMPPRocky replied:
I figured out the issue. It was just getting optimised away again. I've updated the code to use the pointer and now it fails.

XAMPPRocky commented:
@khyperia This should be ready for a second round of review. I have fixed the issue with the test passing, when it should have failed (it was being optimised away). I've also added documentation, and a new `target_env` target feature flag that allows us to specify the target_env we're compiling for, and used to add Vulkan validation to the `push_constant_vulkan.rs` test.

All of the new target feature flag code should be in the second commit if you want to review that separately from the completest changes.

khyperia commented:
> I have fixed the issue with the test passing, when it should have failed (it was being optimised away)

Why was it being optimized away? Changing the test to make it work as expected seems like a workaround to the actual problem, which is that our test setup changed and caused a test to pass that shouldn't have. Changing the test instead of the test setup to make it fail again seems wrong, and may miss something critical that could have changed.

XAMPPRocky commented:
> Why was it being optimized away?

I figured it out. It was because of a different level of `debuginfo`, using cargo we were giving cargo's default (`2` for debug) and this PR was using rustc's default (`0`). I've updated and removed the changes now.

DeanBDean commented on tests/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+compiletest = { version = "0.6.0", package = "compiletest_rs" }
+rustc_codegen_spirv = { path = "../crates/rustc_codegen_spirv" }
+
+# TODO: Needed for handling SPIR-V extension across platforms. Remove once
```
Might be good to add an issue for this, just for visibility?

eddyb commented on tests/ui/lang/core/ptr/allocate_const_scalar.rs:
```diff
@@ -0,0 +1,18 @@
+// Doesn't work, only worked before because I think it got optimized away before
+// hitting the backend.
+
+// Needed because compiletest currently emits `\r\n` on Windows
+// normalize-stderr-test "\r\n" -> "\n"
```
Hmm, how sure are we that this isn't caused by SPIRV-Tools? The Rust side should not be using `\r` at all.

XAMPPRocky replied:
That is a possibility, but I would have assumed `compiletest` would normalise the all of the `stderr` output.

eddyb commented on tests/ui/arch/vector_insert_dynamic.rs:
```diff
@@ -0,0 +1,12 @@
+use spirv_std::arch;
+
+#[spirv(fragment)]
+pub fn main() {
+    let vector = glam::Vec2::new(1.0, 2.0);
+    let expected = glam::Vec2::new(1.0, 3.0);
+    let new_vector = unsafe { arch::vector_insert_dynamic(vector, 1, 3.0) };
+    assert!(new_vector == expected);
```
This seems to be missing the `UVec` case from the original.
**EDIT**: ugh there are actually two "originals".

eddyb commented on tests/ui/arch/vector_extract_dynamic.rs:
```diff
@@ -0,0 +1,11 @@
+use spirv_std::arch;
+
+#[spirv(fragment)]
+pub fn main() {
+    let vector = glam::Vec2::new(1.0, 2.0);
+    let element = unsafe { arch::vector_extract_dynamic(vector, 1) };
+    assert!(2.0 == element);
```
This seems to be missing the `UVec` case from the original.
**EDIT**: ugh there are actually two "originals".

XAMPPRocky commented on Cargo.toml:
```diff
@@ -1,4 +1,5 @@
 [workspace]
+exclude = ["tests/", "target/"]
```
Nope.

XAMPPRocky commented on tests/main.rs:
```diff
@@ -0,0 +1,319 @@
+        std::fs::write(TEST_DEPS_TOML_PATH, CARGO_TOML.as_bytes()).unwrap();
+        std::fs::write(
+            PathBuf::from(TEST_DEPS_PATH).join("src/lib.rs"),
+            TEST_DEPS_SRC,
```
Nope, to successfully build it needs the crate attributes and an entrypoint.

eddyb replied:
Is this because it's a `dylib`? Feels like it should work with an empty `lib.rs` if it were a `lib` crate.

eddyb commented on tests/src/main.rs:
```diff
@@ -0,0 +1,295 @@
+    .any(|o| o.is_none())
+    {
+        clean_project(manifest_dir);
+        build_spirv_std(manifest_dir, codegen_backend_path)
```
Love it :shipit:.

XAMPPRocky commented on tests/ui/lang/core/ptr/allocate_const_scalar.rs:
```diff
@@ -0,0 +1,18 @@
+// Doesn't work, only worked before because I think it got optimized away before
+// hitting the backend.
+
+// Needed because compiletest currently emits `\r\n` on Windows
+// normalize-stderr-test "\r\n" -> "\n"
```
```suggestion
```
